src/basic/user-util.c

```diff
@@ -602,3 +604,27 @@ bool valid_home(const char *p) {
 
         return true;
 }
+
+int maybe_setgroups(size_t size, const gid_t *list) {
+        static int cached_can_setgroups = -1;
+        if (cached_can_setgroups < 0) {
+                _cleanup_free_ char *setgroups_content = NULL;
+                int r;
+
+                r = read_one_line_file("/proc/self/setgroups", &setgroups_content);
+                if (r < 0 && errno != ENOENT)
+                        return r;
+                if (r < 0) {
+                        /* old kernels don't have /proc/self/setgroups, so assume we can use it */
+                        cached_can_setgroups = true;
+                } else {
+                        cached_can_setgroups = !!strcmp(setgroups_content, "deny");
```

Comment:
we try to avoid calling strcmp() if we are not interested in ordering between the strings. If you just want to compare two strings, please use streq(), which is a macro wrapper around it and helps readability.